src/terminal.ts

```diff
@@ -74,9 +76,9 @@ export abstract class Terminal implements ITerminal {
     this._flowControlResume = opt.flowControlResume || FLOW_CONTROL_RESUME;
   }
 
-  protected abstract _write(data: string): void;
+  protected abstract _write(data: string, callback?: (flushed: boolean) => any): void;
 
-  public write(data: string): void {
+  public write(data: string, callback?: (flushed: boolean) => any): void {
```

Comment:
@cancerberoSgx I know this is like 2 years late and you may not care about this anymore but trying to go over all the PRs I've had assigned to me this week and this one slipped through the cracks. You might not need this anymore but I'm onboard with merging something like this in so I merged in master and fixed the conflicts.

I was thinking we should align with how `net.Socket` works where possible but here we've changed `Socket.write`'s return value into a callback which is a little confusing as `Socket.write` also has a callback. What do you think about instead of a callback doing this:

```ts
write(data: string): Promise<boolean>
```

The whether we use promises discussion is resolved at this point and the answer is definitely yes unless it is a very performance critical section to avoid extra overhead (eg. the `onData` event). Since write in a terminal context isn't called that often the small overhead of a Promise is nothing.

I can make these changes myself if you're not interested or strapped for time, but I'm interested in your opinion.